### DISCUSSION TITLE
Enhanced unmanaged pointers (fixed ref)

### DISCUSSION THREAD

**Discussion by DaZombieKiller** [dotnet/csharplang#5366] [Language Ideas]
Currently, the C# language does not have a concept of an unmanaged pointer to constant data (`T const*`). This can lead to complications when wrapping native APIs that need to use them (the `const`-ness information is lost in translation to C#), and when exposing constant unmanaged data (such as a `Guid`.)

I propose exposing a new syntax for pointers: `fixed ref`. A `fixed ref` is equivalent to a normal pointer, but with the ability to be declared `readonly`.

* A `fixed ref T` would map to `T*`
* A `fixed in T` would map to `T const*`
* A `fixed ref readonly T` would map to `T const*`
* A `fixed ref` is considered blittable and can be stored as a field

The compiler would only allow the creation of a `fixed ref` from an `unmanaged` value that is already known to be `fixed`.
* `fixed ref int Ref => ref *pointer;` would be valid
* `fixed ref int Ref => ref classField;` would not be valid

An example where this would be useful is in @tannergooding's TerraFX library. [It currently exposes `readonly` references to constant `Guid` values for COM](https://github.com/terrafx/terrafx.interop.windows/blob/c361364ffb34dee5dab4afeb70ee4c6f53f4b37c/sources/Interop/Windows/um/Unknwnbase/Windows.cs#L61-L82):
```cs
public static ref readonly Guid IID_IClassFactory
{
    get
    {
        ReadOnlySpan<byte> data = new byte[] {
            // snip
        };

        Debug.Assert(data.Length == Unsafe.SizeOf<Guid>());
        return ref Unsafe.As<byte, Guid>(ref MemoryMarshal.GetReference(data));
    }
}
```
Unfortunately, most of the APIs that need to use these values require a `Guid*`, which means you need to use any of the following workarounds:
* `fixed (Guid* iid = &IID_SomeInterface)`
* `Unsafe.AsPointer(ref Unsafe.AsRef(in IID_SomeInterface))`
* A `ref` extension method that returns a pointer using either of the above techniques.
* Copying the `Guid` to a local and retrieving its address with the `&` operator.

The `IID_IClassFactory` example above could alternatively have been declared as `public static Guid* IID_IClassFactory`, but then the `const`-ness information for the value is lost, making it far more prone to misuse. With support for `fixed ref`, it could be changed to:
```cs
public static fixed ref readonly Guid IID_IClassFactory
{
    get
    {
        ReadOnlySpan<byte> data = new byte[] {
            // snip
        };

        Debug.Assert(data.Length == Unsafe.SizeOf<Guid>());
        return ref *(Guid*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(data));
    }
}
```
This can then be safely used with the address-of operator:
```cs
MethodRequiringInterfaceID(&IID_IClassFactory);
```
This also has the additional benefit of being a (source-level) non-breaking change for existing consumers of the API.

**Comment by svick** [comment]
I don't quite understand why introducing a whole new concept would be worthwhile here:

1. If this is about PInvoking APIs that take the equivalent of `Guid*` on the native side, then can't you just declare the method as taking an `in` parameter, or even [`LPStruct`](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices?WT.mc_id=DOP-MVP-5002735#guids) in C#?
2. Why are the workarounds you suggested insufficient?
3. I think adding `const` to pointers would be a much more natural change, would that work here? Or is prohibition of pointer arithmetic necessary?
4. Overall, making it harder to write incorrect code is a good goal, but I feel like this is a fairly complicated feature with very small gains.

**Reply by DaZombieKiller** [reply]
> If this is about PInvoking APIs that take the equivalent of `Guid*` on the native side, then can't you just declare the method as taking an in parameter, or even `LPStruct` in C#?

This is intended for blittable P/Invoke signatures, which can't make use of `ref`/`in` and `[MarshalAs]`.

> Why are the workarounds you suggested insufficient?

They are functional, but extremely verbose and not at all pleasant to use.

> I think adding `const` to pointers would be a much more natural change, would that work here? Or is prohibition of pointer arithmetic necessary?

This is another approach that would work and I would be fine with.

**Reply by SupinePandora43** [reply]
> can't you just declare the method as taking an `in` parameter

DllImportGenerator produces this blittable PInvoke from `NativeMethod(in StrucT @struct)`:
```cs
[GeneratedDllImport]
NativeMethod(in StrucT @struct){
    fixed(StrucT* p = &@struct){
        return NativeMethod__PInvoke(p);
    }
}
[DllImport]
extern NativeMethod__PInvoke(StrucT* @struct);
```

**Comment by alexrp** [comment]
I feel like a proposal for pointer `const`ness has to deal with the fact that C pointers can have `const` at multiple levels, e.g. `const char *const *strings = ...;`.
